storage/kv_test.go

```diff
@@ -821,6 +821,136 @@ func TestWatchableKVWatch(t *testing.T) {
 	}
 }
 
+func TestWatchableKVWatchChanEvent(t *testing.T) {
+	initOp := func(kv WatchableKV) { kv.Put([]byte("foo"), []byte("bar")) }
+	tests := []struct {
+		op   func(kv WatchableKV)
+		wevs []storagepb.Event
+	}{
+		// put event
+		{
+			func(kv WatchableKV) {
+				kv.Put([]byte("foo2"), []byte("bar"))
+			},
+			[]storagepb.Event{
+				{
+					Type: storagepb.PUT,
+					Kv:   &storagepb.KeyValue{Key: []byte("foo2"), Value: []byte("bar"), CreateRevision: 2, ModRevision: 2, Version: 1},
+				},
+			},
+		},
+		// delete event
+		{
+			func(kv WatchableKV) {
+				kv.DeleteRange([]byte("foo"), nil)
+			},
+			[]storagepb.Event{
+				{
+					Type: storagepb.DELETE,
+					Kv:   &storagepb.KeyValue{Key: []byte("foo")},
+				},
+			},
+		},
+		// put event in txn
+		{
+			func(kv WatchableKV) {
+				id := kv.TxnBegin()
+				kv.TxnPut(id, []byte("foo2"), []byte("bar"))
+				kv.TxnEnd(id)
+			},
+			[]storagepb.Event{
+				{
+					Type: storagepb.PUT,
+					Kv:   &storagepb.KeyValue{Key: []byte("foo2"), Value: []byte("bar"), CreateRevision: 2, ModRevision: 2, Version: 1},
+				},
+			},
+		},
+		// delete event in txn
+		{
+			func(kv WatchableKV) {
+				id := kv.TxnBegin()
+				kv.TxnDeleteRange(id, []byte("foo"), nil)
+				kv.TxnEnd(id)
+			},
+			[]storagepb.Event{
+				{
+					Type: storagepb.DELETE,
+					Kv:   &storagepb.KeyValue{Key: []byte("foo")},
+				},
+			},
+		},
+		// delete non-existant event
+		{
+			func(kv WatchableKV) {
+				kv.DeleteRange([]byte("foo2"), nil)
+			},
+			[]storagepb.Event{},
+		},
+		// multiple events in txn
+		{
+			func(kv WatchableKV) {
+				id := kv.TxnBegin()
+				kv.TxnPut(id, []byte("foo2"), []byte("bar"))
+				kv.TxnDeleteRange(id, []byte("foo"), nil)
+				kv.TxnEnd(id)
+			},
+			[]storagepb.Event{
+				{
+					Type: storagepb.PUT,
+					Kv:   &storagepb.KeyValue{Key: []byte("foo2"), Value: []byte("bar"), CreateRevision: 2, ModRevision: 2, Version: 1},
+				},
+				{
+					Type: storagepb.DELETE,
+					Kv:   &storagepb.KeyValue{Key: []byte("foo")},
+				},
+			},
+		},
+		// put twice on one key in txn, only last one should be watched
+		{
+			func(kv WatchableKV) {
+				id := kv.TxnBegin()
+				kv.TxnPut(id, []byte("foo"), []byte("bar2"))
+				kv.TxnPut(id, []byte("foo"), []byte("bar3"))
+				kv.TxnEnd(id)
+			},
+			[]storagepb.Event{
+				{
+					Type: storagepb.PUT,
+					Kv:   &storagepb.KeyValue{Key: []byte("foo"), Value: []byte("bar3"), CreateRevision: 1, ModRevision: 2, Version: 3},
+				},
+			},
+		},
+		// TODO: put and delete one new key in txn, no events should be watched
```

Comment:
xiang suggests an possible alternative that watcher should not hide events on the same key in txn. I will think about this more.